### PR TITLE
SpringDescription parameter for the AnimationController fling method

### PR DESCRIPTION
**Pull request by nt4f04uNd**
## Description

I provided the fling method with a new parameter `springDescription`. It just adds more customization to the methon by allowing to override the existing internal constant variable and so permitting developers to manipulate how the resulting fling animation will look like.

## Related Issues

#64400 - the original issue, though I then decided that tolerance shouldn't be public.

## Tests

Added

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Comment by LongCatIsLooong** [comment]
I'm still not sure why `animeWith` won't work if you want to use a custom simulation. As I understand it, the "direction" of the animation has little meaning for a custom simulation. Could you shed a little light on why having the direction set to be always forward is problematic?

**Comment by nt4f04uNd** [comment]
@LongCatIsLooong refer to my comment here https://github.com/flutter/flutter/issues/60419#issuecomment-662664171.

[See where](https://youtu.be/XDkt_BDiJyE) I want to use this. When I have such animation, I want to have all the valid animation statues: forward, complete, reverse, dismissed. Using `animateWith` will give me only two of them, which is so bad, I mean it's so wrong when I have animation that actually goes back, but status says me that it's forward instead. This is severe bug to me.

**Comment by LongCatIsLooong** [comment]
@nt4f04uNd that makes sense. Thanks for the explanation! In that case the documentation needs to be updated, as the custom spring is not guaranteed to be critically damped. Also I think the interaction between underdamped/overdamped springs with `lowerBound`, `upBound` is worth documenting.

**Comment by goderbauer** [comment]
@nt4f04uNd Do you have plans to follow up on the feedback soon-ish?

**Comment by nt4f04uNd** [comment]
@LongCatIsLooong @goderbauer sorry for delayed updates. I updated docs a bit, but I guess I'm not competent enough to document the relationships between various types of springs and `lowerBound` and `upperBound`. 

I tested with different bounds and different types of springs, and `criticallyDamped` and `overDamped` springs behave as I would expect from them - which is overdamped is mostly like critically damped but slowed down, though the `underDamped` springs behave weirdly - they bounce way too fast and it looks more like flickering, rather than actual bouncing. 

I'm not that much into all that stuff, so if you think it should be documented more detailed, I ask you to do that instead of me

**Comment by LongCatIsLooong** [comment]
@nt4f04uNd yeah it gets a bit hairy when the given spring has a damping ratio that's not 1. Does your use case involve using a non-critically damped spring? If not I think we can add an assert to ensure the custom `SpringDescriptor` supplied describes a critically damped spring and add in the doc that this method currently only supports critically damped springs.

**Comment by nt4f04uNd** [comment]
@LongCatIsLooong over damped seems to be working pretty ok and I use it as I want to slow down a bit the default critically damped spring and make it a bit smoother (I use the one I included in tests). 

Did you observe any problems with over damped springs, because I experience problems only with under damped as I said and if we want to add asserts, I would do that allowing to use both critically and over damped springs.

**Comment by LongCatIsLooong** [comment]
I think the problem with using an overdamped `SpringDescription` is that it doesn't naturally stop at `lowerBound` (or `upperBound`), this could be unintuitive in your use case too because the animation may stop prematurely (e.g., if you present a new sheet with that fling animation, the sheet presentation animation may stop half screen if you set the damping ratio to a large number).

**Comment by nt4f04uNd** [comment]
@LongCatIsLooong that's not the case. If you set ratio to very large number, the animation will always end up in the upper/lower bound (depending on the direction), but what you actually see when you do this - is drastically slowed down movement, it's not stopped. You can set `timeDilation` to e.g. `0.001` to ensure that.

But I think maybe we should actually document in this method the behaviour of different spring types to not confuse anyone who will actually try to do that.

**Comment by nt4f04uNd** [comment]
@LongCatIsLooong also, just to be sure, as with large ratio, the animation may start moving just because of applying dilation (I'm not aware how it internally works), so I added listener to the controller and it's reports about changes even without dilation, even though on the screen there's no animation at all

**Comment by LongCatIsLooong** [comment]
Oh you're right, it should eventually converge to 0:
https://github.com/flutter/flutter/blob/1510865fb45513124601a587229543c488103b14/packages/flutter/lib/src/physics/spring_simulation.dart#L232-L235

> But I think maybe we should actually document in this method the behaviour of different spring types to not confuse anyone who will actually try to do that.

Sounds good :+1:

**Comment by nt4f04uNd** [comment]
@LongCatIsLooong will we restrict `underDamped` though?

**Comment by LongCatIsLooong** [comment]
It looks like we clamp the `value` to [lowerBound, upperBound] so I assume with an underdamped spring it should still oscillate except when the sheet is supposed to overshoot, it stays at `lowerBound` or `upperBound`  instead?
https://github.com/flutter/flutter/blob/1510865fb45513124601a587229543c488103b14/packages/flutter/lib/src/animation/animation_controller.dart#L785-L798

I don't know if there's a use case for that, the movement doesn't feel natural (as the animation suddenly stops when it goes out of range, losing all its momentum, then regains the momentum after a short while). But I'm OK with just documenting the behavior and warning developers about the weirdness.  @goderbauer do you think we should disallow underdamped springs in this case?

**Comment by goderbauer** [comment]
(PR triage): @LongCatIsLooong @nt4f04uNd are there still plans for this PR?

**Comment by nt4f04uNd** [comment]
@goderbauer I've been planning to add a few comment lines about how the springs work, but been waiting for your response

> But I'm OK with just documenting the behavior and warning developers about the weirdness. @goderbauer do you think we should disallow underdamped springs in this case?

File: packages/flutter/test/animation/animation_controller_test.dart [hunk 1]
```diff
@@ -255,6 +255,29 @@ void main() {
...
+        springDescription: SpringDescription.withDampingRatio(
+            mass: 0.01,
+            stiffness: 10.0,
+            ratio: 2.0,
+        ),
+      );
+      tick(const Duration(milliseconds: 0));
+      tick(const Duration(milliseconds: 50));
+
+      expect(customSpringController.value < controller.value, true);
```
Comment (goderbauer) [inline packages/flutter/test/animation/animation_controller_test.dart:278]: nit: looks like everything from line 260 to 278 is indented too much.

**Comment by goderbauer** [comment]
> @goderbauer do you think we should disallow underdamped springs in this case?

I think we should disallow it if it doesn't make sense and from the discussion so far it sounds like it wouldn't really make sense? We can always revisit that if we have a use case that requires it.

**Comment by LongCatIsLooong** [comment]
Thanks! There seems to be a merge conflict, could you resolve it?

**Comment by nt4f04uNd** [comment]
Turns out I can't

**Comment by nt4f04uNd** [comment]
done

File: packages/flutter/lib/src/animation/animation_controller.dart [hunk 2]
```diff
@@ -645,18 +645,30 @@ class AnimationController extends Animation<double>
...
-  /// and [upperBound]) and initial velocity.
+  /// Drives the animation with a spring (within [lowerBound] and [upperBound])
+  /// and initial velocity.
   ///
   /// If velocity is positive, the animation will complete, otherwise it will
   /// dismiss.
   ///
+  /// The default spring that is driven by this method is critically damped,
+  /// but the [springDescription] can be used to drive by a custom spring.
+  /// With this parameter you can make the fling slower, or otherwise faster.
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/animation/animation_controller.dart:656]: nit: I feel "fling slower" or "fling faster" is a bit inaccurate as it also depends on the initial velocity. How about something like:

The [springDescription] parameter can be used to specify a custom [SpringType.criticallyDamped] or [SpringType.overDamped] spring to drive the animation with. Defaults to null,  which uses a [SpringType.criticallyDamped] spring. See [SpringDescription.withDampingRatio] for how to create a suitable [SpringDescription].

**Review by LongCatIsLooong — APPROVED** [review]
Thank you for the PR! LGTM.

**Comment by nt4f04uNd** [comment]
looks nice

**Comment by LongCatIsLooong** [comment]
@nt4f04uNd could you do a `git pull upstream/master` and see if that fixes the CI failure?

**Comment by nt4f04uNd** [comment]
@LongCatIsLooong do you mean `git pull upstream master`, yup? So, I should do that on my pull request branch and then what, commit something?

**Comment by LongCatIsLooong** [comment]
@nt4f04uNd yes. The linux docs check on CI seems to be failing for some unrelated reason so let's see if syncing to head fixes that.

**Pull request merged**